File: sim/core/bulk/statistics.go
package bulk

import (
	"cmp"
	"container/heap"
	"math"
	"slices"

	"github.com/wowsims/tbc/sim/core/proto"
)

type bulkSimResultMinHeap []*BulkSimCandidateResult

func (h bulkSimResultMinHeap) Len() int { return len(h) }

func (h bulkSimResultMinHeap) Less(i, j int) bool {
	return h[i].DpsMetrics.Avg < h[j].DpsMetrics.Avg
}

func (h bulkSimResultMinHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }

func (h *bulkSimResultMinHeap) Push(x any) {
	*h = append(*h, x.(*BulkSimCandidateResult))
}

func (h *bulkSimResultMinHeap) Pop() any {
	old := *h
	n := len(old)
	result := old[n-1]
	*h = old[:n-1]
	return result
}

// Converts a target relative error into an iteration count using standard
// error: stdev / sqrt(iterations). The combination multiplier is a practical
// multiple-candidate adjustment so large candidate sets use more iterations
// without paying the full Bonferroni cost.
func getBulkSimTargetIterations(targetErrorPct float64, metrics *proto.DistributionMetrics, candidateCount int) int32 {
	if metrics == nil || metrics.Avg <= 0 {
		return 0
	}

	targetError := metrics.Avg * (targetErrorPct / 100)
	if targetError <= 0 {
		return 0
	}

	combinationMultiplier := bulkSimCombinationErrorMultiplier(candidateCount)
	return int32(math.Ceil(math.Pow((metrics.Stdev*combinationMultiplier)/targetError, 2)))
}

// Keeps candidates that could still plausibly be the best result after
// accounting for sim variance. The top MinSurvivors by mean are always
// retained, then any candidate whose upper interval overlaps the best
// candidate's lower interval is kept. A soft cap prevents pathological stages
// from forwarding the entire candidate set when many results are tied.
func selectBulkSimSurvivors(results []*BulkSimCandidateResult, baseline *BulkSimCandidateResult, iterations int32, config BulkSimStageConfig, originalCandidateCount int) []BulkSimCandidate {
	maxSurvivors := getBulkSimStageMaxSurvivors(config, len(results))
	if maxSurvivors == 0 || len(results) <= maxSurvivors {
		return bulkSimResultsToCandidates(results)
	}

	bestMetrics := baseline.DpsMetrics
	for _, result := range results {
		if result == nil || result.DpsMetrics == nil {
			continue
		}
		if result.DpsMetrics.Avg > bestMetrics.Avg {
			bestMetrics = result.DpsMetrics
		}
	}
	intervalMultiplier := bulkSimSurvivorIntervalMultiplier(originalCandidateCount, config.CullingCoefficient)
	bestLowerBound := bestMetrics.Avg - bulkSimDpsError(bestMetrics, iterations)*intervalMultiplier

	meanSurvivors := topBulkSimResults(results, config.MinSurvivors)
	survivors := make([]*BulkSimCandidateResult, 0, maxSurvivors)
	seen := make(map[int32]bool)
	for _, result := range meanSurvivors {
		survivors = append(survivors, result)
		seen[result.Candidate.Index] = true
	}
	for _, result := range results {
		if result == nil || result.DpsMetrics == nil || seen[result.Candidate.Index] {
			continue
		}

		if bulkSimCandidateIsCulled(result.DpsMetrics, bestMetrics, bestLowerBound, iterations, intervalMultiplier) {
			continue
		}
		survivors = append(survivors, result)
		seen[result.Candidate.Index] = true
	}

	softMaxSurvivors := maxSurvivors * bulkSimSurvivorSoftCapMultiplier
	if len(survivors) > softMaxSurvivors {
		survivors = topBulkSimResults(survivors, softMaxSurvivors)
	}

	return bulkSimResultsToCandidates(survivors)
}

// The marginal cull test compares the sum of both candidates' half-widths, which is
// sqrt(2) wider than the standard error of the difference it is really testing. Keeping
// that factor when switching to the paired estimate means pairing only removes the
// variance the shared seeds already cancel - it does not additionally tighten the
// interval.
const bulkSimPairedIntervalConservatism = math.Sqrt2

// Decides whether a candidate is far enough behind the leader to drop out.
//
// Every candidate is simmed on the same seed sequence, so most of the per-iteration
// noise is shared between any two of them. Comparing marginal standard errors throws
// that away; differencing the paired per-iteration values keeps it, which resolves the
// same gap with fewer iterations. Falls back to the marginal comparison when the values
// are unavailable or not aligned (different iteration counts).
func bulkSimCandidateIsCulled(metrics *proto.DistributionMetrics, bestMetrics *proto.DistributionMetrics, bestLowerBound float64, iterations int32, intervalMultiplier float64) bool {
	if pairedError, ok := bulkSimPairedDpsError(metrics, bestMetrics); ok {
		return bestMetrics.Avg-metrics.Avg > pairedError*intervalMultiplier*bulkSimPairedIntervalConservatism
	}

	candidateUpperBound := metrics.Avg + bulkSimDpsError(metrics, iterations)*intervalMultiplier
	return candidateUpperBound < bestLowerBound
}

// Standard error of the mean per-iteration difference between a candidate and the
// leader. Reports false when the two runs cannot be paired; a zero error is a valid
// result - it means the candidate trailed the leader by the same amount every
// iteration, which is the strongest evidence pairing can give.
func bulkSimPairedDpsError(metrics *proto.DistributionMetrics, bestMetrics *proto.DistributionMetrics) (float64, bool) {
	if metrics == nil || bestMetrics == nil {
		return 0, false
	}
	values, bestValues := metrics.AllValues, bestMetrics.AllValues
	if len(values) == 0 || len(values) != len(bestValues) {
		return 0, false
	}

	var sum, sumSq float64
	for idx, value := range values {
		difference := value - bestValues[idx]
		sum += difference
		sumSq += difference * difference
	}
	count := float64(len(values))
	mean := sum / count
	variance := math.Max(0, sumSq/count-mean*mean)
	return math.Sqrt(variance / count), true
}

func bulkSimDpsError(metrics *proto.DistributionMetrics, iterations int32) float64 {
	if metrics == nil || iterations <= 0 {
		return 0
	}
	return metrics.Stdev / math.Sqrt(float64(iterations))
}

// Intentionally much lighter than a strict Bonferroni correction. Bulk Sim
// needs to avoid false culls among many candidates, but absolute proof of the
// full ordering would require infeasible iteration counts for near-tied gear
// sets.
func bulkSimCombinationErrorMultiplier(candidateCount int) float64 {
	return math.Sqrt(math.Max(1, math.Log10(math.Max(float64(candidateCount), bulkSimCombinationLogMin))))
}

func bulkSimSurvivorIntervalMultiplier(candidateCount int, cullingCoefficient float64) float64 {
	return cullingCoefficient * bulkSimCombinationErrorMultiplier(candidateCount)
}

func bulkSimObservedErrorPct(metrics *proto.DistributionMetrics, iterations int32, candidateCount int) float64 {
	if metrics == nil || metrics.Avg <= 0 || iterations <= 0 {
		return 0
	}
	return bulkSimDpsError(metrics, iterations) * bulkSimCombinationErrorMultiplier(candidateCount) / metrics.Avg * 100
}

// Reports the worst relative error across the baseline and every candidate.
// Using the max is intentionally conservative: one noisy candidate can still
// affect culling or final top-result confidence.
func bulkSimObservedStageErrorPct(baseline *BulkSimCandidateResult, results []*BulkSimCandidateResult, iterations int32, candidateCount int) float64 {
	observedErrorPct := 0.0
	if baseline != nil {
		observedErrorPct = bulkSimObservedErrorPct(baseline.DpsMetrics, iterations, candidateCount)
	}
	for _, result := range results {
		if result == nil {
			continue
		}
		observedErrorPct = math.Max(observedErrorPct, bulkSimObservedErrorPct(result.DpsMetrics, iterations, candidateCount))
	}
	return observedErrorPct
}

func getBulkSimStageTargetIterations(targetErrorPct float64, baseline *BulkSimCandidateResult, results []*BulkSimCandidateResult, candidateCount int) int32 {
	targetIterations := int32(0)
	if baseline != nil {
		targetIterations = max(targetIterations, getBulkSimTargetIterations(targetErrorPct, baseline.DpsMetrics, candidateCount))
	}
	for _, result := range results {
		if result == nil {
			continue
		}
		targetIterations = max(targetIterations, getBulkSimTargetIterations(targetErrorPct, result.DpsMetrics, candidateCount))
	}
	return targetIterations
}

func topBulkSimResults(results []*BulkSimCandidateResult, limit int) []*BulkSimCandidateResult {
	if limit <= 0 || len(results) == 0 {
		return nil
	}
	// Both branches sort on DpsMetrics.Avg, so both drop entries that cannot supply one.
	// The filter is inline rather than a shared pre-pass so the heap branch below keeps
	// streaming over results within its bounded `limit` allocation.
	if len(results) <= limit {
		topResults := make([]*BulkSimCandidateResult, 0, len(results))
		for _, result := range results {
			if result == nil || result.DpsMetrics == nil {
				continue
			}
			topResults = append(topResults, result)
		}
		sortBulkSimResultsByDps(topResults)
		return topResults
	}

	topResults := make(bulkSimResultMinHeap, 0, limit)
	for _, result := range results {
		if result == nil || result.DpsMetrics == nil {
			continue
		}
		if topResults.Len() < limit {
			heap.Push(&topResults, result)
			continue
		}
		if result.DpsMetrics.Avg > topResults[0].DpsMetrics.Avg {
			topResults[0] = result
			heap.Fix(&topResults, 0)
		}
	}

	result := []*BulkSimCandidateResult(topResults)
	sortBulkSimResultsByDps(result)
	return result
}

func sortBulkSimResultsByDps(results []*BulkSimCandidateResult) {
	slices.SortFunc(results, func(a, b *BulkSimCandidateResult) int {
		return cmp.Compare(b.DpsMetrics.Avg, a.DpsMetrics.Avg)
	})
}

func bestBulkSimDps(results []*BulkSimCandidateResult) float64 {
	best := 0.0
	for _, result := range results {
		if result != nil && result.DpsMetrics != nil {
			best = math.Max(best, result.DpsMetrics.Avg)
		}
	}
	return best
}

File: ui/core/wasm/bulk_sim/statistics.ts
import { DistributionMetrics } from '../../proto/api';
import { Z_95 } from '../../utils';
import { BULK_SIM_COMBINATION_LOG_MIN, BULK_SIM_CULLING_COEFFICIENT, BULK_SIM_SURVIVOR_SOFT_CAP_MULTIPLIER } from './constants_auto_gen';
import { getBulkSimStageMaxSurvivors } from './stage';
import { ConcurrentBulkSimCandidate, ConcurrentBulkSimCandidateResult, ConcurrentBulkSimStageConfig } from './types';

export const getBulkSimTargetIterations = (targetErrorPct: number, metrics: DistributionMetrics | undefined, candidateCount: number): number => {
	if (!metrics || metrics.avg <= 0) return 0;

	const targetError = metrics.avg * (targetErrorPct / 100);
	if (targetError <= 0) return 0;

	const combinationMultiplier = bulkSimCombinationErrorMultiplier(candidateCount);
	return Math.ceil(Math.pow((metrics.stdev * combinationMultiplier) / targetError, 2));
};

// The marginal test compared the SUM of both half-widths, which is sqrt(2) wider than the
// difference standard error it was really testing. Keeping that factor means pairing only
// removes the variance the shared seeds genuinely cancel, rather than also tightening the rule.
const BULK_SIM_PAIRED_INTERVAL_CONSERVATISM = Math.SQRT2;

// Standard error of the mean per-iteration difference between a candidate and the
// leader. Returns undefined when the two runs cannot be paired; zero is a valid result -
// it means the candidate trailed the leader by the same amount every iteration, which is
// the strongest evidence pairing can give.
export const bulkSimPairedDpsError = (metrics: DistributionMetrics | undefined, bestMetrics: DistributionMetrics | undefined): number | undefined => {
	const values = metrics?.allValues;
	const bestValues = bestMetrics?.allValues;
	if (!values || !bestValues || values.length === 0 || values.length !== bestValues.length) return undefined;

	let sum = 0;
	let sumSq = 0;
	for (let idx = 0; idx < values.length; idx++) {
		const difference = values[idx] - bestValues[idx];
		sum += difference;
		sumSq += difference * difference;
	}
	const count = values.length;
	const mean = sum / count;
	return Math.sqrt(Math.max(0, sumSq / count - mean * mean) / count);
};

// Decides whether a candidate is far enough behind the leader to drop out. Every candidate
// is simmed on the same seed sequence, so most of the per-iteration noise is shared between
// any two of them; differencing the paired values keeps that, resolving the same gap with
// fewer iterations. Falls back to the marginal comparison when the runs cannot be paired.
const bulkSimCandidateIsCulled = (
	metrics: DistributionMetrics,
	bestMetrics: DistributionMetrics | undefined,
	bestLowerBound: number,
	iterations: number,
	intervalMultiplier: number,
): boolean => {
	const pairedError = bulkSimPairedDpsError(metrics, bestMetrics);
	if (pairedError !== undefined && bestMetrics) {
		return bestMetrics.avg - metrics.avg > pairedError * intervalMultiplier * BULK_SIM_PAIRED_INTERVAL_CONSERVATISM;
	}

	const candidateUpperBound = metrics.avg + bulkSimDpsError(metrics, iterations) * intervalMultiplier;
	return candidateUpperBound < bestLowerBound;
};

const bulkSimDpsError = (metrics: DistributionMetrics | undefined, iterations: number): number => {
	if (!metrics || iterations <= 0) return 0;
	return metrics.stdev / Math.sqrt(iterations);
};

const bulkSimCombinationErrorMultiplier = (candidateCount: number): number =>
	Math.sqrt(Math.max(1, Math.log10(Math.max(candidateCount, BULK_SIM_COMBINATION_LOG_MIN))));

const bulkSimSurvivorIntervalMultiplier = (candidateCount: number, cullingCoefficient: number): number =>
	cullingCoefficient * bulkSimCombinationErrorMultiplier(candidateCount);

const bulkSimObservedErrorPct = (metrics: DistributionMetrics | undefined, iterations: number, candidateCount: number): number => {
	if (!metrics || metrics.avg <= 0 || iterations <= 0) return 0;
	return (bulkSimDpsError(metrics, iterations) * bulkSimCombinationErrorMultiplier(candidateCount) * 100) / metrics.avg;
};

export const bulkSimObservedStageErrorPct = (
	baseline: ConcurrentBulkSimCandidateResult | undefined,
	results: ConcurrentBulkSimCandidateResult[],
	iterations: number,
	candidateCount: number,
): number => {
	let observedErrorPct = bulkSimObservedErrorPct(baseline?.dpsMetrics, iterations, candidateCount);
	for (const result of results) {
		observedErrorPct = Math.max(observedErrorPct, bulkSimObservedErrorPct(result.dpsMetrics, iterations, candidateCount));
	}
	return observedErrorPct;
};

export const getBulkSimStageTargetIterations = (
	targetErrorPct: number,
	baseline: ConcurrentBulkSimCandidateResult | undefined,
	results: ConcurrentBulkSimCandidateResult[],
	candidateCount: number,
): number => {
	let targetIterations = getBulkSimTargetIterations(targetErrorPct, baseline?.dpsMetrics, candidateCount);
	for (const result of results) {
		targetIterations = Math.max(targetIterations, getBulkSimTargetIterations(targetErrorPct, result.dpsMetrics, candidateCount));
	}
	return targetIterations;
};

export const topBulkSimResults = (results: ConcurrentBulkSimCandidateResult[], limit: number): ConcurrentBulkSimCandidateResult[] => {
	if (limit <= 0 || results.length == 0) return [];
	return results
		.filter(result => result.dpsMetrics)
		.sort((a, b) => b.dpsMetrics!.avg - a.dpsMetrics!.avg)
		.slice(0, limit);
};

// originalCandidateCount is the candidate count for the WHOLE run, not the surviving count at this
// stage: the culling interval must be widened for every candidate ever compared, otherwise later
// stages test against a narrower interval than the multiple-comparison correction assumes.
export const selectBulkSimSurvivors = (
	results: ConcurrentBulkSimCandidateResult[],
	baseline: ConcurrentBulkSimCandidateResult,
	iterations: number,
	config: ConcurrentBulkSimStageConfig,
	originalCandidateCount: number,
): ConcurrentBulkSimCandidate[] => {
	const maxSurvivors = getBulkSimStageMaxSurvivors(config, results.length);
	if (maxSurvivors === undefined || results.length <= maxSurvivors) {
		return results.map(result => result.candidate);
	}

	let bestMetrics = baseline.dpsMetrics;
	for (const result of results) {
		if (result.dpsMetrics && (!bestMetrics || result.dpsMetrics.avg > bestMetrics.avg)) {
			bestMetrics = result.dpsMetrics;
		}
	}

	const intervalMultiplier = bulkSimSurvivorIntervalMultiplier(originalCandidateCount, config.cullingCoefficient ?? BULK_SIM_CULLING_COEFFICIENT);
	const bestLowerBound = (bestMetrics?.avg ?? 0) - bulkSimDpsError(bestMetrics, iterations) * intervalMultiplier;
	const meanSurvivors = topBulkSimResults(results, config.minSurvivors ?? 0);
	let survivors = meanSurvivors.slice();
	const seen = new Set(survivors.map(result => result.candidate.index));

	for (const result of results) {
		if (!result.dpsMetrics || seen.has(result.candidate.index)) continue;

		if (bulkSimCandidateIsCulled(result.dpsMetrics, bestMetrics, bestLowerBound, iterations, intervalMultiplier)) continue;

		survivors.push(result);
		seen.add(result.candidate.index);
	}

	const softMaxSurvivors = maxSurvivors * BULK_SIM_SURVIVOR_SOFT_CAP_MULTIPLIER;
	if (survivors.length > softMaxSurvivors) {
		survivors = topBulkSimResults(survivors, softMaxSurvivors);
	}

	return survivors.map(result => result.candidate);
};

// Whether any adjacent pair in the DPS-sorted finalist list is still statistically
// unresolved: paired z inside the same Z_95 threshold the UI's significance test uses.
// The cull's BULK_SIM_PAIRED_INTERVAL_CONSERVATISM factor is deliberately NOT applied -
// it compensates for the cull test's sum-of-half-widths formulation, while this is a
// plain two-sided test on the paired difference. An unpairable pair (or one with zero
// paired error) cannot be improved by more lockstep iterations and counts as resolved.
export const bulkSimUnresolvedFinalistPair = (sortedFinalists: ConcurrentBulkSimCandidateResult[]): boolean => {
	for (let idx = 0; idx + 1 < sortedFinalists.length; idx++) {
		const upper = sortedFinalists[idx];
		const lower = sortedFinalists[idx + 1];
		const pairedError = bulkSimPairedDpsError(lower.dpsMetrics, upper.dpsMetrics);
		if (pairedError === undefined || pairedError === 0) continue;
		if (Math.abs((upper.dpsMetrics?.avg ?? 0) - (lower.dpsMetrics?.avg ?? 0)) <= Z_95 * pairedError) return true;
	}
	return false;
};
